docs: add mfa guide (#9353)
supabase/supabase: apps/reference/docs/guides/auth/auth-mfa.mdx

---
id: auth-mfa
title: 'Multi-Factor Authentication'
description: Add an additional layer of security to your apps with Supabase Auth multi-factor authentication.
---

import { Mermaid } from 'mdx-mermaid/Mermaid';

:::note
Multi-Factor Authentication is in early access preview only.
:::

> Multi-factor authentication (MFA) or sometimes also called two-factor
> authentication (2FA) is a way to add an additional layer of security to your
> application by requiring users to verify their identity through one or more
> additional verification steps.

Reviewer: ```suggestion
Multi-factor authentication (MFA), sometimes called two-factor
authentication (2FA), adds an additional layer of security to your
application by requiring users to verify their identity through
additional verification steps.
```

> It is considered a best practice for all but the simplest of applications.

Reviewer: ```suggestion
It is considered a best practice to use MFA for your applications.
```

Reviewer: I think "best practice" implies "some exceptions"? Perhaps we don't need to add the type of exceptions

> Users who use weak passwords or have compromised social login accounts are
> prone to malicious account takeovers. These can be prevented by using
> multi-factor authentication because they require the user to provide proof of
> both of these:

Reviewer: ```suggestion
Users with weak passwords or compromised social login accounts are
prone to malicious account takeovers. These can be prevented with
MFA because they require the user to provide proof of
both of these:
```

- Something they know.  
Password, or access to a social-login account.
- Something they have.  
Access to an authenticator app (a.k.a. TOTP), mobile phone or recovery code.

## Overview

> Supabase Auth implements only TOTP multi-factor authentication. This type of

Reviewer 2: /nit should we define TOTP? Think the adjective might be more well known then the full term though
```suggestion
Supabase Auth implements only Time-based One Time Factor(TOTP) multi-factor authentication. This type of
```

multi-factor authentication uses a timed one-time password generated from an
authenticator app in the control of users.

> You need to prepare your application for two important flows when adding MFA.

Reviewer: ```suggestion
Applications using MFA require two important flows:
```

groups of users.

Once users have enrolled or signed-in with a factor, Supabase Auth adds
additional metadata to the user's access token (JWT) that your application can
use to allow or deny access.

This information is represented by an [Authenticator Assurance
Level](https://pages.nist.gov/800-63-3-Implementation-Resources/63B/AAL/), a
standard measure about the assurance Supabase Auth has of the user's identity
for that particular session. There are two levels recognized today:

1. **Assurance Level 1: `aal1`**  
Means that the user's identity was verified using a conventional login method
such as email+password, magic link, one-time password, phone auth or social
login.
2. **Assurance Level 2: `aal2`**  
Means that the user's identity was additionally verified using at least one
second factor, such as a TOTP code.

This assurance level is encoded in the `aal` claim in the JWT associated with
the user. By decoding this value you can create custom authorization rules in
your frontend, backend and database that will enforce the MFA policy that works
for your application. JWTs without an `aal` claim are at the `aal1` level.

## Adding to your app

> Adding multi-factor authentication to your app involves these three steps:

Reviewer 2: /nit can use MFA since it seems like we're using MFA in the previous paragraphs
```suggestion
Adding MFA to your app involves these three steps:
```

1. **Add enrollment flow.**  
You need to provide a UI within your app that your users will be able to set-up
MFA in. You can add this right after sign-up, or as part of a separate flow in
the settings portion of your app.
2. **Add challenge step to login.**  
If a user has set-up MFA, your app's login flow needs to present a challenge
screen to the user asking them to prove they have access to the additional
factor.
3. **Enforce rules for MFA logins.**  
Once your users have a way to enroll and log in with MFA, you need to enforce
authorization rules across your app: on the frontend, backend, API servers or
Row-Level Security policies.

### Add enrollment flow

Before users can secure their accounts with MFA, you need to provide them with
a UI where they can set up their additional authentication factor. We call this
the enrollment flow.

Most applications add the enrollment flow in two places within their app:

> 1. Right after login or sign up.  
> This lets users quickly set up MFA immediately after they log in or create an
> account. We recommend encouraging all users to set up MFA if that makes sense
> for your application. Many applications offer this as an opt-in step in an
> effort to reduce onboarding friction.
> 2. From within a settings page.  
> Allows users to set up, disable or modify their MFA settings.

Reviewer 3: I think we should leave out the recommendation and leave it up to the dev. We can likely assume they're already encouraging MFA for their users if they're going through this guide. This is just a suggestion though—will leave it up to you!
```suggestion
1. Right after login or sign up.  
This lets users set up MFA immediately after they log in or create an
account. Many applications offer this as an opt-in step in an
effort to reduce onboarding friction.
2. From within a settings page.  
Allows users to set up, disable, or modify their MFA settings.
```

Author: I actually would like to leave it as is. It lets users figure out exactly what they can do with MFA, and that there are some tricky choices to be made. It then also follows the recommendations with the RLS policies below...

We recommend building one generic flow that you can reuse in both cases with
minor modifications.

Enrolling a factor for use with MFA takes three steps:

> 1. Calling the `supabase.auth.mfa.enroll()` API.  

Reviewer 3: Couple nits (and definitely correct me if you disagree):
- I think the "API" in this case is the Auth/MFA API
- We should refer to these calls as the API methods/endpoints but we don't have to say that explicitly. For example, step 1 can just be:
```suggestion
1. Call `supabase.auth.mfa.enroll()`.  
```

> The output of this call is a QR code and a secret. You should display the QR
> code to the user and ask them to scan it with their authenticator application.
> If they are unable to scan, you can show them the secret in plain text which
> they can type up or paste into the app.

Reviewer 3: ```suggestion
This method returns a QR code and a secret. Display the QR
code to the user and ask them to scan it with their authenticator application.
If they are unable to scan the QR code, show the secret in plain text which
they can type or paste into their authenticator app.
```

> 2. Calling the `supabase.auth.mfa.challenge()` API.  

Reviewer 2: I like that we're highlighting this explicitly as two separate steps

- `onCancelled` is a callback that notifies the other components that the user
  has clicked the `Cancel` button.

```tsx
function EnrollMFA({
  onEnrolled,
  onCancelled,
}: {
  onEnrolled: () => void;
  onCancelled: () => void;
}) {
  const [factorId, setFactorId] = useState("");
  const [qr, setQR] = useState(""); // holds the QR code image SVG
  const [verifyCode, setVerifyCode] = useState(""); // contains the code entered by the user
  const [error, setError] = useState(""); // holds an error message

  const onEnableClicked = () => {
    setError("");

    (async () => {
      const challenge = await supabase.auth.mfa.challenge({ factorId });
      if (challenge.error) {
        setError(challenge.error.message);
        throw challenge.error;
      }

      const challengeId = challenge.data.id;

      const verify = await supabase.auth.mfa.verify({
        factorId,
        challengeId,
        code: verifyCode,
      });
      if (verify.error) {
        setError(verify.error.message);
        throw verify.error;
      }

      onEnrolled();
    })();
  };

  useEffect(() => {
    (async () => {
      const { data, error } = await supabase.auth.mfa.enroll({

>         factorType: "TOTP",

Reviewer 2: ```suggestion
        factorType: "totp",
```

/nit This should be lowercase
